### PR TITLE
Update tile and decompose attention pass

### PR DESCRIPTION
**Pull request by harsh-nod**
This patch adds the following updates to the tile and decompose pass:
1. Does matmuls in fp16 with fp32 accumulators
2. Does softmax in fp32
3. Reduces number of divisions to 1 (from 2) for faster performance
4. Change exp to exp2 for faster performance

### CODE REVIEW

**Comment by harsh-nod** [comment]
@MaheshRavishankar @hanhanW - please can you review these changes?

**Comment by GMNGeoffrey** [comment]
I have a consistent repro for the memory error that surfaced here: https://github.com/openxla/iree/issues/13967

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp [hunk 1]
```diff
@@ -176,33 +194,17 @@ extractSlices(Value key, Value value, Value query, Value output,
...
-  sizes[2] = headDimension;
-  offsets[0] = ivs[0];
-  Value updatedAcc = builder.create<tensor::InsertSliceOp>(
-      loc, newResult, result, offsets, sizes, strides);
-  offsets = SmallVector<OpFoldResult>(queryShape.size() - 1, zero);
-  sizes = SmallVector<OpFoldResult>{one, sequenceTileLength};
-  strides = SmallVector<OpFoldResult>(queryShape.size() - 1, one);
-  return std::make_tuple(updatedAcc, newMax, newSum);
+  return std::make_tuple(newResult, newMax, newSum);
 }
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp]: Do we still need this function...? It looks very redundant.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp [hunk 2]
```diff
@@ -245,24 +266,67 @@ createAttentionBody(Value keySlice, Value valueSlice, Value querySlice,
...
 
+static Value extractOrInsertOutputSlice(Value src, Value dst,
+                                        ArrayRef<int64_t> queryShape,
+                                        OpFoldResult sequenceTileLength,
+                                        OpFoldResult headDimension,
+                                        Location loc, OpBuilder &builder) {
+  auto one = builder.getIndexAttr(1);
+  auto zero = builder.getIndexAttr(0);
+  SmallVector<OpFoldResult> strides(3, one);
+  SmallVector<OpFoldResult> sizes{one, sequenceTileLength, headDimension};
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp]: Prefer using assignment-like with `=` when initializing with the intended literal value. It let us have uniform style, and avoid bugs like below:

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.
```

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp (hunk 1 above)
Comment (harsh-nod) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp]: Thanks, will remove this function.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp (hunk 2 above)
Comment (harsh-nod) [reply llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/TileAndDecomposeAttention.cpp]: Sure, will remove the braces initializer. Thanks!

**Comment by harsh-nod** [comment]
@hanhanW - sorry this patch has gotten buried as I was out for a while. Could you please take a look at this again?

**Review by hanhanW — APPROVED** [review]

**Comment by harsh-nod** [comment]
Thanks!

**Pull request merged**